### PR TITLE
fix for /tmp/tmp access issues

### LINKED ISSUE #7218: Compiler not found: /tmp/tmp bug
```
$ pip install numba
Collecting numba
  Downloading numba-0.53.1-cp39-cp39-manylinux2014_x86_64.whl (3.4 MB)
     |████████████████████████████████| 3.4 MB 1.2 MB/s 
Collecting numpy>=1.15
  Downloading numpy-1.21.0-cp39-cp39-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.7 MB)
     |████████████████████████████████| 15.7 MB 1.1 MB/s 
Requirement already satisfied: setuptools in /usr/lib/python3/dist-packages (from numba) (52.0.0)
Collecting llvmlite<0.37,>=0.36.0rc1
  Downloading llvmlite-0.36.0-cp39-cp39-manylinux2010_x86_64.whl (25.3 MB)
     |████████████████████████████████| 25.3 MB 829 kB/s 
Installing collected packages: numpy, llvmlite, numba
  WARNING: The scripts f2py, f2py3 and f2py3.9 are installed in '/home/olaf/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed llvmlite-0.36.0 numba-0.53.1 numpy-1.21.0

$ ./t.py 
          
$ ./t.py 

$ chmod 000 /tmp/tmp

$ ./t.py            
Traceback (most recent call last):
  File "/home/olaf/./t.py", line 5, in <module>
    cc = CC('interpolation')
  File "/home/olaf/.local/lib/python3.9/site-packages/numba/pycc/cc.py", line 65, in __init__
    self._toolchain = Toolchain()
  File "/home/olaf/.local/lib/python3.9/site-packages/numba/pycc/platform.py", line 78, in __init__
    self._raise_external_compiler_error()
  File "/home/olaf/.local/lib/python3.9/site-packages/numba/pycc/platform.py", line 121, in _raise_external_compiler_error
    raise RuntimeError(msg)
RuntimeError: Attempted to compile AOT function without the compiler used by `numpy.distutils` present. If using conda try:

#> conda install gcc_linux-64 gxx_linux-64

$ cat t.py 
#!/usr/bin/python3

from numba.pycc import CC

cc = CC('interpolation')

# cc.compile()

$ pip list|grep numba
numba            0.53.1
```

### PR DESCRIPTION
**Pull request by ChiCheng45**
Changes are made on how the temp files and temp dirs are created by numba for testing the compilers in the function _check_external_compiler in platform.py

This is my proposed fix for the issue #7218 which can occur when multiple users are working on a shared resource such as on a HPC.  Basically if one user builds numba AOT code they will create a directory named /tmp/tmp which may mean that other users will unable to build AOT due to a permission error.

fixes #7218

### CODE REVIEW

**Comment by gmarkall** [comment]
@ChiCheng45 Many thanks for the PR! Is this now ready for review?

**Comment by ChiCheng45** [comment]
@gmarkall Hi yes this is ready for review now but I'm not sure why this docs are taking so long to be built, has it gotten stuck?

**Comment by gmarkall** [comment]
> @gmarkall Hi yes this is ready for review now but I'm not sure why this docs are taking so long to be built, has it gotten stuck?

Many thanks, I've added it to the queue. I didn't see a problem with the doc building - I think sometimes RTD builders wait in a queue for a little while though.

File: numba/pycc/platform.py [hunk 1]
```diff
@@ -26,40 +28,28 @@ def get_configs(arg):
...
-                ntf.write(simple_c)
-                ntf.flush()
-                ntf.close()
-                # *output_dir* is set to avoid the compiler putting temp files
-                # in the current directory.
-                compiler.compile([ntf.name], output_dir=gettempdir())
-        except Exception: # likely CompileError or file system issue
+            # *output_dir* is set to avoid the compiler putting temp files
+            # in the current directory.
+            compiler.compile([ntf.name], output_dir=Path(ntf.name).anchor)
```
Comment (apmasell) [inline numba/pycc/platform.py:47]: This is the only change in behaviour. Could you please revert all the other changes except this part?

Comment (ChiCheng45) [reply numba/pycc/platform.py]: ok should be done now.

**Review by apmasell — DISMISSED** [review]

**Comment by sklam** [comment]
BFID `numba_smoketest_cpu_yaml_116`

File: numba/pycc/platform.py [hunk 2]
```diff
@@ -57,7 +58,7 @@ def _check_external_compiler():
                 ntf.close()
                 # *output_dir* is set to avoid the compiler putting temp files
                 # in the current directory.
-                compiler.compile([ntf.name], output_dir=gettempdir())
```
Comment (stuartarchibald) [inline numba/pycc/platform.py:60]: Does removing this mean there's no need to import `gettempdir` now?

Comment (ChiCheng45) [reply numba/pycc/platform.py:60]: doesn't look like it, I've removed unused imports in my latest commit

**Comment by OlafvdSpek** [comment]
Would this run with `output_dir='/'` on Linux?

**Comment by ChiCheng45** [comment]
@OlafvdSpek yea that should also work but I preferred Path(ntf.name).anchor since they might have the tmp dir on another drive on windows

**Comment by OlafvdSpek** [comment]
Yeah, was just asking what the effective output_dir would be. Wouldn't this create a new `c:\\tmp` dir on Windows?

**Comment by ChiCheng45** [comment]
@OlafvdSpek no it creates a directory in whatever the temp directory is for example, C:\Users\username\AppData\Local\Temp\tmp4kygk7kx.

**Pull request merged**